### ISSUE TITLE
Changeset Spec Exceeds Database Limit

### ISSUE THREAD

**Issue by Piszmog** [sourcegraph/sourcegraph#39242]
The table `changeset_specs` has a column `spec` of type `jsonb`. PostgreSQL limits `jsonb` to 255 MB. If a changeset exceeds this limit, the following error occurs,

```text
ERROR: string too long to represent as jsonb string (SQLSTATE 54000)
```

## Possible Fixes

* Prevent and changesets from exceeding the 255 MB limit by adding "guards"
    * If a changeset surpasses the limit, maybe point the user to docs to decrease changeset sizes (for example, [transformchanges](https://docs.sourcegraph.com/batch_changes/references/batch_spec_yaml_reference#transformchanges))
* Rethink storing the `spec` in the `jsonb` column in the `chageset_specs` table.
    * This will probably be a larger effort and may need a follow up Issue for long term 

## Scenarios

Logging GraphQL requests impacts the behavior of the errors (simple `502` vs actual error message). But, the backing error is the same.

### Environment Variables

* `NO_GRAPHQL_LOG=true`

#### Batch Spec

```yaml
# Generates 676.3 MB changeset spec
name: large-changeset
description: Add Hello World to READMEs

on:
  - repositoriesMatchingQuery: file:README.md repo:^github.com/odoo/odoo$

steps:
  - run: find . -name "*.po" -exec sed -i 's/b/c/g' {} \;
    container: alpine:3

changesetTemplate:
  title: Hello World
  body: My first batch change!
  branch: hello-world-1 # Push the commit to this branch.
  commit:
    message: Append Hello World to all README.md files
  published: false # Do not publish any changes to the code hosts yet
```

#### Response

```json
{
  "message": "inserter.Flush: ERROR: string too long to represent as jsonb string (SQLSTATE 54000)",
  "path": [
    "createChangesetSpec"
  ]
}
```

#### Logs

```text
[       frontend] ERROR Inserter.database.batch.Flush batch/batch.go:224 operation.error {"batchSize": 14, "payloadSize": 24, "tableName": "changeset_specs", "columnNames": "rand_id,spec,batch_spec_id,repo_id,user_id,diff_stat_added,diff_stat_changed,diff_stat_deleted,created_at,updated_at,fork_namespace,external_id,head_ref,title", "numColumns": 14, "maxBatchSize": 32760, "count": 1, "elapsed": 4.235910458, "error": "ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
[       frontend] ERROR enterprise.batches.dbstore.CreateChangesetSpec store/changeset_specs.go:128 operation.error {"Count": 1, "count": 1, "elapsed": 32.111408542, "error": "inserter.Flush: ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
[       frontend] ERROR enterprise.batches.service.CreateChangesetSpec service/service.go:758 operation.error {"count": 1, "elapsed": 61.167864, "error": "inserter.Flush: ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
```

### Environment Variables

N/A

#### Batch Spec

```yaml
# Generates 1.0 GB changeset spec
name: large-changeset
description: Add Hello World to READMEs

on:
  - repositoriesMatchingQuery: file:README.md repo:^github.com/odoo/odoo$

steps:
  - run: find . -name "*.po" -exec sed -i 's/a/b/g' {} \;
    container: alpine:3

changesetTemplate:
  title: Hello World
  body: My first batch change!
  branch: hello-world-1 # Push the commit to this branch.
  commit:
    message: Append Hello World to all README.md files
  published: false # Do not publish any changes to the code hosts yet
```

#### Response

```text
error: 502 Bad Gateway
```

#### Logs

```text
[       frontend] ERROR Inserter.database.batch.Flush batch/batch.go:224 operation.error {"batchSize": 14, "payloadSize": 24, "tableName": "changeset_specs", "columnNames": "rand_id,spec,batch_spec_id,repo_id,user_id,diff_stat_added,diff_stat_changed,diff_stat_deleted,created_at,updated_at,fork_namespace,external_id,head_ref,title", "numColumns": 14, "maxBatchSize": 32760, "count": 1, "elapsed": 4.927489625, "error": "ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
[       frontend] ERROR enterprise.batches.dbstore.CreateChangesetSpec store/changeset_specs.go:128 operation.error {"Count": 1, "count": 1, "elapsed": 32.148294167, "error": "inserter.Flush: ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
[       frontend] ERROR enterprise.batches.service.CreateChangesetSpec service/service.go:758 operation.error {"count": 1, "elapsed": 63.868031875, "error": "inserter.Flush: ERROR: string too long to represent as jsonb string (SQLSTATE 54000)"}
```

**Comment by eseliger** [comment]
Interesting, another data point that storing patches in a json column isn't .. great 😬

**Comment by LawnGnome** [comment]
> * Prevent and changesets from exceeding the 255 MB limit by adding "guards"
>   
>   * If a changeset surpasses the limit, maybe point the user to docs to decrease changeset sizes (for example, [transformchanges](https://docs.sourcegraph.com/batch_changes/references/batch_spec_yaml_reference#transformchanges))
> 
> * Rethink storing the `spec` in the `jsonb` column in the `chageset_specs` table.
>   
>   * This will probably be a larger effort and may need a follow up Issue for long term

I kind of suspect we might end up doing both, since the first one _should_ be a quick win and the second would be a better long term solution.

**Comment by chrispine** [comment]
is the changeset data compressed in the json? If not, that could at least make this more rare (but still doesn't solve the problem)

How rarely does this happen?

**Comment by LawnGnome** [comment]
Sadly, we can't compress the data in situ because we use PostgreSQL's JSON operators to index and introspect into the data structure.

I think we should split this: putting guards in is probably 0.5d and could be a P0/1; denormalising the data we need into other fields and then storing the JSON as a (possibly compressed) blob is probably (wild guess) 2d and backlog.

**Comment by chrispine** [comment]
As part of this, create a followup ticket in Sustaining for the rest of the work Erik lays out above

**Comment by eseliger** [comment]
Landed a PR that raises this limit for the diff field to the limit of the bytea column type, which is 1G. This will be available from 4.0 on.
